### PR TITLE
[Docs] Document the filtering module

### PR DESCRIPTION
**Pull request by timgrein**
## Part of https://github.com/elastic/enterprise-search-team/issues/4017

Add documentation for the most important classes and methods in the filtering module. Skipped on the documentation, where it doesn't add any information, which can not be derived from the class name.

### CODE REVIEW

File: connectors/filtering/basic_rule.py [hunk 1]
```diff
@@ -94,6 +95,8 @@ def try_coerce(value):
 
 
 class RuleMatchStats:
+    """Class to record how many documents a basic rule matched and which policy it used."""
```
[the remark this answers is not in the document]
Comment (artem-shelkovnikov) [reply connectors/filtering/basic_rule.py:98]: Also looking at the class, why is it made immutable? I see it creates a new instance of RuleMatchStats every time something is added to the stats?

Why a simple hash would not make this job?

File: connectors/filtering/basic_rule.py [hunk 2]
```diff
@@ -114,13 +117,22 @@ def __eq__(self, other):
 
 
 class BasicRuleEngine:
+    """Class executing a list of basic rules in order against a document."""
```
Comment (artem-shelkovnikov) [inline connectors/filtering/basic_rule.py:120]: Same as above - comment can explain better:

1. What does it actually do - current description is so abstract it does not explain what the class is supposed to be used for
2. Provide an example of usage

File: connectors/filtering/basic_rule.py [hunk 3]
```diff
@@ -114,13 +117,22 @@ def __eq__(self, other):
...
     def __init__(self, rules):
         self.rules = rules
         self.rules_match_stats = {
             BasicRule.DEFAULT_RULE_ID: RuleMatchStats(Policy.INCLUDE, 0)
         }
 
     def should_ingest(self, document):
+        """Method to check, whether a document should be ingested or not.
+
+        On default the document will be ingested, if it doesn't match any rule.
```
Comment (artem-shelkovnikov) [inline connectors/filtering/basic_rule.py:131]:
```suggestion
        By default the document will be ingested, if it doesn't match any rule.
```

What about "not by default" case?

File: connectors/filtering/basic_rule.py [hunk 4]
```diff
@@ -220,6 +232,8 @@ def from_string(cls, string):
 
 
 class BasicRule:
+    """Class representing a basic rule, which is part of the connector's filtering."""
```
Comment (artem-shelkovnikov) [inline connectors/filtering/basic_rule.py:235]: Same as above - "Class" s redundant, what is this class for, why is it a basic rule and what's not a basic rule?

File: connectors/filtering/basic_rule.py [hunk 5]
```diff
@@ -253,6 +267,23 @@ def default_rule(cls):
...
+        A basic rule matches or doesn't match a document based on the following comparisons:
+            - STARTS_WITH: Does the document's field value start with the basic rule's value?
+            - ENDS_WITH: Does the document's field value end with the basic rule's value?
+            - CONTAINS: Does the document's field value contain the basic rule's value?
+            - REGEX: Does the document's field value match the basic rule's regex?
+            - LESS_THAN: Is the document's field value less than the basic rule's value?
+            - GREATER_THAN: Is the document's field value greater than the basic rule's value?
+            - EQUALS: Is the document's field value equal to the basic rule's value?
+
+        If the basic rule is the default rule it's always a match.
```
Comment (artem-shelkovnikov) [inline connectors/filtering/basic_rule.py:281]: What's a "default rule"?

File: connectors/filtering/basic_rule.py [hunk 6]
```diff
@@ -253,6 +267,23 @@ def default_rule(cls):
...
+        If the field is not in the document it's always a no match.
+
+        Arguments:
+        - `document`: document to check, if it matches
```
Comment (artem-shelkovnikov) [inline connectors/filtering/basic_rule.py:285]:
```suggestion
        - `document`: document to check for matches
```

File: connectors/filtering/basic_rule.py [hunk 7]
```diff
@@ -287,6 +318,14 @@ def is_include(self):
         return self.policy == Policy.INCLUDE
 
     def coerce_rule_value_based_on_document_value(self, doc_value):
+        """Method coercing the value inside the basic rule.
+
+        This method tries to coerce the value inside the basic rule to the type used in the document.
+
+        Arguments:
+        - `doc_value`: one value of a document
```
Comment (artem-shelkovnikov) [inline connectors/filtering/basic_rule.py:326]:
```suggestion
        - `doc_value`: value of the field in the document to coerce
```

File: connectors/filtering/validation.py [hunk 8]
```diff
@@ -196,6 +196,12 @@ def validate(cls, rules):
 
 
 class BasicRulesSetSemanticValidator(BasicRulesSetValidator):
+    """Class validating, that a set of rules does not contain semantic duplicates.
```
Comment (artem-shelkovnikov) [inline connectors/filtering/validation.py:199]:
```suggestion
    """BasicRulesSetSemanticValidator can be used to validate that a set of filtering rules does not contain semantic duplicates.
```

File: connectors/filtering/validation.py [hunk 9]
```diff
@@ -196,6 +196,12 @@ def validate(cls, rules):
...
+
+    A semantic duplicate is defined as two basic rules having the same values for `field`, `rule` and `value`.
+    Therefore, two basic rules are also seen as semantic duplicates, if their `policy` values differ.
```
Comment (artem-shelkovnikov) [inline connectors/filtering/validation.py:202]: What happens in this case? Is the rule set declared as invalid?

File: connectors/filtering/validation.py [hunk 10]
```diff
@@ -255,6 +261,8 @@ def validate(cls, rule):
 
 
 class BasicRuleNoMatchAllRegexValidator(BasicRuleValidator):
+    """Class validating that a basic rule does not use a match all regex."""
```
Comment (artem-shelkovnikov) [inline connectors/filtering/validation.py:264]:
```suggestion
    """BasicRuleNoMatchAllRegexValidator can be used to check that a basic rule does not use a match all regex."""
```

What's the deal with match all regex here?

File: connectors/filtering/validation.py [hunk 11]
```diff
@@ -279,6 +287,8 @@ def validate(cls, basic_rule_json):
 
 
 class BasicRuleAgainstSchemaValidator(BasicRuleValidator):
+    """Class validating if a basic rule follows the specified json schema."""
```
Comment (artem-shelkovnikov) [inline connectors/filtering/validation.py:290]:
```suggestion
    """BasicRuleAgainstSchemaValidator can be used to check if basic rule follows specified json schema."""
```

File: connectors/filtering/validation.py (hunk 10 above)
Comment (timgrein) [reply connectors/filtering/validation.py:264]: I actually can not recall, what was the exact rationale to prohibit match all regular expressions. Should be somewhere in the original implementation doc.

File: connectors/filtering/basic_rule.py (hunk 6 above)
Comment (timgrein) [reply connectors/filtering/basic_rule.py:285]: A document has one match per rule or no match. "for matches" would imply that the doc can be matched multiple for the same rule IMO

File: connectors/filtering/basic_rule.py (hunk 5 above)
Comment (timgrein) [reply connectors/filtering/basic_rule.py]: Added an explanation

File: connectors/filtering/validation.py (hunk 9 above)
Comment (timgrein) [reply connectors/filtering/validation.py:202]: Added an explanation.

**Review by tarekziade — APPROVED** [review]

**Pull request merged**